### DISCUSSION TITLE
"Locked Time Period Always Used" doesn't appear to ever be active

### DISCUSSION THREAD

**Discussion by trallen** [NightscoutFoundation/xDrip#3460] [Bug reports] [answered]
I prefer to have my graph set to display the last three hours, but even having long-pressed the 3 hour button to lock the time view, as suggested in the documentation, I find that the graph always resets to the default period when for example, opening the app.

I had set "Locked Time Period Always Used" to see if that would help, but it appears to make no difference. 

Digging through the code, and particularly commit 4e868c4, it looks like the active piece should be "home-locked-hours", to set which_hours in Home.java. However, I can't see anywhere where home-locked-hours is set.

There's a strong possibility that I'm following a wild goose chase on this; my main aim is to get the graph to have a consistent time scale, so that I can compare slopes easily.

Thank you!

**Comment by Navid200** [comment]
It works for me.  However, there is an exception and I wonder if that may be causing a confusion.
Please note that if you can see the forecast low dotted red line on the graph, the time axis will be expanded so that you can see when you will be low.

I admit that we can reduce the expansion to the time you cross your low setting rather than an additional hour as is now.  But, there are so many other things that have kept me busy for now.

Would you please tell me if you see the issue even if you cannot see the dotted red line?
I have shown the dotted red line in the graphs on this page: https://navid200.github.io/xDrip/docs/Alerts/ForecastLow.html

**Reply by trallen** [reply]
Thanks, that's very helpful.

I disabled the Forecast Lows feature, but it had no effect (I'm not going low currently! I imagine if I was, I'd see the behaviour you describe). But then I looked at the Predictive Simulations setting, and when I turned that off, the graph no longer resized itself.

Is there any way to get Forecast Lows and Predictive Simulations working while still not having the graph resize itself? (Perhaps by shifting the start of the displayed time window, rather than by resizing the scale to fit a longer time period.)

If not, I guess I can live without the (very useful) predictive simulations. 

Thanks!

**Reply by Navid200** [reply]
I am open to suggestions.

A forecast shows a curve in the future.    If you choose to see something in the future without extending the time axis accordingly, how could you see anything in the future?!

**Reply by trallen** [reply]
I was thinking the display could shift the window. 

So, say, without the forecasts, the window displays the last three hours: 14:51 to 17:51.

A forecast is added, covering an hour. Rather than displaying 14:51 to 18:51, it would display 15:51 to 18:51 on screen.

Of course, there's a caveat: if the forecast is for longer than 3 hours. I'm not sure that's possible, but one alternative is to not show the full forecast on screen; in the example above, the main window would show 15:21 to 18:21, and the user can scroll backwards to the left as per normal, or forwards to the right to show the remaining half hour that isn't displayed.

**Reply by Navid200** [reply, marked as answer]
I personally have no objection to this.
We currently have more serious issues with the display.  For example, the start/stop flags sometimes cause the y axis to extend dramatically.  
It's just that there are so many things that need to be fixed and some of us are working on them.  But, we don't have enough developers.

The time axis behavior may be the way it is for a reason I am not aware of.  So, that means I cannot make any promises.
But, please click on "Create issue from discussion" in the right pane.
It will convert this to an issue.  I suggest that you choose a title for the issue that best captures your suggestion.

**Reply by trallen** [reply]
Rather than create more work for the developers, I'll create a pull request when I have some free time. Any advice you have off the top of your head about the best to implement this (or any summaries about, or links to the issues with the display) would be useful and appreciated!

**Reply by Navid200** [reply]
That would be awesome.
This talks about how we work although I had intended to add a lot more: https://navid200.github.io/xDrip/docs/dev/dev.html

This open issue shows the start/stop flag problem: https://github.com/NightscoutFoundation/xDrip/issues/3081

This one is about changing the penalty time from 24 hours to something shorter.  https://github.com/NightscoutFoundation/xDrip/issues/2173

You can see all the open issues.  They all need to be addressed.  Some are related to display.

**Reply by trallen** [reply]
Thanks, that's a great starting point. I'll revive my local repo from the last time I worked on xDrip! :-)

**Reply by jonprogers** [reply]
Adding to the list of things that impact the x scaling the green decay curve for IoB does too.
I have very slow insulin absorption so run with a much longer DIA than anyone would regard as normal so each bolus compresses the x-axis to an extent to show the full decay of each bolus dose.
I don't see this as an issue as I get a quick reminder of when I'll be free of any bolus insulin whenever I look at it although the displayed bg values look very compressed to the left so it looked strange until I appreciated the value of what I was seeing.
I look forward to seeing if I can gain even more insight from any new balance you find for the time axis scaling!!

**Reply by trallen** [reply]
@jonprogers Please review the pull request at #3464 to see if this meets your needs, or at least, doesn't negatively affect your use?

**Reply by jonprogers** [reply]
Nothing proposed in #3464, would have any negative impact on me, thanks for asking. Looks like a really adaptable solution to the issue identified. Good luck getting it merged!
